### ISSUE TITLE
Supress printing of line numbers

### ISSUE THREAD

**Issue by forthrin** [beyondgrep/ack3#311]
ack v3.4.0 always prints line numbers:

    123: Hello, World!

Is there an option to not print line numbers, eg.

    Hello, World!

Curiously, the help and man pages say nothing of this fundamental functionality.

#142 might relate to this, but nothing happening there for a couple of years.

Does this mean we are stuck with line numbers?

**Comment by petdance** [comment]
Can you give a specific command line that you're having problems with?

Does the `-h` not do what you want?

**Comment by forthrin** [comment]
`-h` surpresses both filenames and line numbers (though the latter is not mentioned in the man pages). I want to keep filenames, but surpress line numbers.

       -h, --no-filename
           Suppress the prefixing of filenames on output when multiple files
           are searched.

**Comment by petdance** [comment]
If I'm understanding, instead of 

```
t/text/amontillado.txt
61:"And yet some fools will have it that his taste is a match for your
98:the foot of the descent, and stood together on the damp ground of
156:"A huge human foot d'or, in a field azure; the foot crushes a serpent

t/text/raven.txt
93:Swung by seraphim whose foot-falls tinkled on the tufted floor.
```

you want:

```
t/text/amontillado.txt
"And yet some fools will have it that his taste is a match for your
the foot of the descent, and stood together on the damp ground of
"A huge human foot d'or, in a field azure; the foot crushes a serpent

t/text/raven.txt
Swung by seraphim whose foot-falls tinkled on the tufted floor.
```

Why is that? What's the use case? Not arguing, just trying to understand.

**Comment by forthrin** [comment]
Exactly. I think it was where the length of the line number string varied from 1 to 4 characters, thereby skewing the text content on the lines accordingly, making it difficult to relate them to each other. Suppressing line numbers would seem like a one-liner to implement, though I fully support developer inertia to prevent feature bloat. An alternative would be to right-align the line numbers, eg. `printf("%'*d")`, though I think suppressing line numbers seems like the proper fix with many potential benefits.

**Comment by n1vux** [comment]
Knowing how many digits to allow `%*d` either requires scanning the whole file before output *or* a heuristic for the largest sensible filesize in lines when the inode only saves filesize in bytes (not even in characters, but bytes).  

Not an impossible heuristic: `log10($filelength_bytes)` is a firm, generous, but not crisp upper bound on number of digits in a linenumber 
(assuming we don't count `VT` as some multiple of `CRLF`s !!).
(A crisp heuristic that is never wrong will be hard.)

**Comment by n1vux** [comment]
it may be all the way back in ack1.$large that I last spoke in favor of `--[no-]line-number` for full control of output, since i don't see my handle on comments on the referenced Ack3.0 ticket nor it's Ack2 precursor.

I support full control of line numbering separable from filenames 
since there are times that I do want filenames but not line numbers, 
and having filenames once as headers rather than inline as `:` delimited fields is often preferable (but of course not always) 
so the obvious piping to `cut -d: -f` won't do what i want either.

**Comment by n1vux** [comment]
I will note a weird workaround that does work - putting the `cut` into the ack `pager` option avoids interlining the filename:

`ack get_file_id --pager='cut -d: -f2|less -iR'`

Not intuitive but it works!

(This workaround could be included _pro tempore_ in Cookbook https://github.com/beyondgrep/website/issues/103 )

**Comment by parv** [comment]
Sometimes I find line numbers to be distracting, especially when reading the search result with context option (e.g -C 4). Not having line numbers (along with ":" & "-" prefixes to lines) would help much.

**Comment by TheUnixDude** [comment]
I see this feature has been an ask for about 15 years now.

The use case is: I have two log files - the same log file on two different hosts. Call them file1 and file2. I need to search them for a particular expected event.
```
hostA$ ack EVENT file1 > results1
hostB$ ack EVENT file2 > results2
```
The events should be "the same" within each of file1 and file2 (thru fileN) - the same order, in the same events. But there could be any N number of intervening entries thus the line numbers will never match.

A diff of results1 and results2 should be empty. But it's not, because `ack` corrupts the files by inserting data which is not in either file - the line numbers.

**Comment by forthrin** [comment]
To those who might be interested, I switched to `rg` (since 15 weeks, maybe), after reading the astoundingly altruistic comparison table below. It has some other thorns (legacy UNIX option names that have been around since the epoch, break you do not!), but you tend to forgo that when you're grepping past your theoretical disk read speed.

https://beyondgrep.com/feature-comparison/

**Comment by petdance** [comment]
> after reading the astoundingly altruistic comparison table below.

I appreciate you appreciating that. 

I don't want people to use ack. I want people to use whatever tool best fits their needs. This isn't about my ego to be able to say "More people use mine than yours."

As far as I'm concerned, any project that "competes" (I term I dislike here) with others should have a similar comparison chart.

I just wish mine wasn't so out of date because ripgrep has some other features that aren't reflected on the chart.

**Comment by n1vux** [comment]
For @TheUnixDude  ,  the `-h` option mentioned previously is usable to suppress line-numbers in _your_ use-case. 

`-h` will _normally_ suppress both filename and line numbers in output, but with only _one_ file being processed by each of two ack commands, ack omits the filename anyway, so in this case, `-h` omits line-numbers. 

(For cases where one wants one or more file-names out of many possible files **and** matching lines **but** without line-numbers from each, it is **not** sufficient, and we do not have **that**, which is what the existing feature-requests have been requesting. Eventually I'll probably make a PR for it but ... TUITs.)

Also FWIW, note that in many (?most?) modern shells, you can do a diff of two ack or whatever commands without temporary files via so-called "Process Substitution" (which is using named pipes under the hood):
 `diff <(ack -hw EVENT $fileA) <(ack -hw EVENT $fileB)`. 
When order isn't guaranteed, a useful alternative idiom is  
`comm -3 <(sort $fileA) <(sort $fileB)` 
.

**Comment by n1vux** [comment]
Andy wishes to be fairer to e.g. RipGrep 

> I just wish [my comparison chart] wasn't so out of date because ripgrep has some other features that aren't reflected on the chart.

Tracking multiple moving targets in such a comparison is hard. Harder than it looks!  ❤  

For those following along at home — Adding new features to the chart is harder than it looks because we can't just add one awesome feature for just ripgrep, to be fair we have to check if any of the other covered apps have it too.
Or adding another App to the table, we have to comb it's docs for all the possible features, as well as see if it will actually fit on reasonable screen sizes. Just because it fits on _my_ Monitor of Unusual Size isn't good enough!  (And so many folks are reading the web on 6" iPhones in Portrait mode now, the exact opposite. CSS is supposed to solve this but ... it doesn't quite.)
(Oh, and probably none of us are on the release announce list of all the BeyondGrep listed apps, so we don't even know immediately that there's new features.)

The Comparison chart can never be perfect; good enough has to be good enough.

(I'd say ~~patches~~ PRs welcome but as I said, patching the comparison table is harder than it looks, so probably not so much, really. 😢  )

**Comment by forthrin** [comment]
I wrote a minimal grep using a Ruby script. It takes but a dozen lines or two to get 80% of what you use most of the time working. The only tactile drawback was performance (for huge/many files), but did no optimization either.

I suspect the reason why a project like ack gets complex to maintain and add to is either:

A) Written in C or similar language where string handling is a nightmare
B) Very particular features that require token parsing or other advanced techniques

Neither seems to be true, as the whole utility is 3000 lines of perl.

If the reason is, C) Eternal toilet roll facing direction debates with users, I don't wanna know about it :P

Have you considered rewriting from scratch, porting to a terser or more flexible language, etc.?

**Comment by n1vux** [comment]
> Have you considered rewriting from scratch, porting to a terser or more flexible language, etc.?

in addition to our having considered a total rewrite, it's been done (by others, with other priorities). 
I'd say see the **Other grep-like tools** section on **More Tools** page of project homepage [BeyondGrep.com>more-tools](https://beyondgrep.com/more-tools/), but since you @forthrin switched to *rg* already and tagged the [feature chart](https://beyondgrep.com/feature-comparison/) , **you** don't need that information.

ack is the way it is because it scratches Andy's itch and I help because it mostly scratches mine too. 

Why? 
While other *-grep in C or Rust or ... can offer `-P perlre` using the `pcre2` library, a perl-based *-grep such as **ack** gives us Perl's (irregular) ^Regular Expressions^ that are 100% compatible with our current installed Perl, satisfying the principle of Least Surprise.

(re the actual ticket: I'd like to implement this request, as it bugs me too, but I never get around to it either. Since the use-cases where this is most obviously useful are unstructured data and natural language, they're outside Andy's declared scope for Ack, a Programmer's `eeegrep -P` for code, so I'm not pushing.)

**Comment by forthrin** [comment]
Should probably check out the alternatives one day. Yes, `rg` is quick, almost as the maintainer is quick to be condescending to users. And once the initial good vibe is tainted like that, it's hard to get it back.

One would have thought by now that the romancing of the 90s BOFH was but a embarrassing relic, and even the most hardened down-voters, self-declared Tech Gods and other online Comic Book Guy type figures had gotten sufficiently self-aware by public discourse, intervention-induced psych evaluation, or just their moms telling them to straighten up or else spend the rest of your Christmases alone.

But those habits really die hard with some. The only that can get away with it is Linus. That's it. Everyone else has to be nice.

**Comment by forthrin** [comment]
Pending five years so closing.

**Comment by forthrin** [comment]
I'm outta here.

**Comment by petdance** [comment]
I can open a new ticket if you'd prefer. I still think is a useful idea.

**Comment by forthrin** [comment]
Unsubscribed from the issue because five years is long enough. But sure @ me if you ever get around to it.
